[major] Add public modules (#130)
chipsalliance/firrtl-spec: spec.md

## Circuits

A FIRRTL circuit is a named collection of FIRRTL modules.  Each module is a
hardware "unit" that has ports, registers, wires, and may instantiate other
modules (see: [@sec:modules]).  (This is the same concept as a Verilog

> `module`{.verilog}.)  A public module is usable outside of the current circuit.

Reviewer: what does it mean to be "usable". Should we say that a public module follows the defined verilog ABI? And that it blocks constant propagation across its boundaries? I'm still unclear if that is what "public" means.

Reviewer: i guess you clarify this below

Author: Talking about what it means to be "public" is difficult here as we haven't even introduced module, yet. This is more of rewriting the spec to put things in a more logical order. The requirements of `public` are specified in the "Public Modules" subsection of the "Module" section which follows the "Circuit" section. @mmaloney-sf has brought up the fact that this whole thing is hard to read and it may make sense to reverse all these sections. I.e., introduce `module` before `circuit`. This type of thing is largely out of scope of this PR. I agree we should do some big rewrites, though.

Consider the following circuit named `Foo`.  This contains two modules, `Bar`
and `Baz`.  Module `Baz` is marked public.

``` firrtl
circuit Foo :
  module Bar :
    input a: UInt<1>
    output b: UInt<1>

    connect b, a

>   public module Bar :

Reviewer 2: ```suggestion
  public module Baz :
```

>   public module Bar :

Reviewer: did you want one of these to be Baz?

Can you show that you can have another `public module SomethingElse` that can also instantiate the non-public module?

    input a: UInt<1>
    output b: UInt<1>

    inst bar of Bar
    connect bar.a, a
    connect b, bar.b
```

> A circuit may contain public modules that instantiate no common instances.

Reviewer 2: Saying same thing you said a few different ways but just to be clear:

Dropping "main module".  Circuits are kinda like packages of public entirely disjoint circuits.
Is that right?

1) Can they instantiate the same extmodule's? Or is the rule about not sharing any instances only for definitions within the circuit?
2) Can they share anything at all, like the recent top-level definitions (class/extclass, properties, type alias, so on)?
3) Other than the filelist detail, I think what's proposed here is very much like what you'd get with different circuits (each with own "top") but instead in same file-- with difference being moving multiple circuits into same circuit requires them to be disjoint (where-as today multiple circuits can of course extmodule and use the other).  IS that about right?

Reviewer: i think you read the eats shoots and leaves like i did ... the statement is that they are *allowed* to instantiate the same (non-public) modules and extmodules. They do not HAVE to share any commmon modules and extmodules.

In other words, (https://mentor.ieee.org/myproject/Public/mytools/draft/styleman.pdf)

- public modules `MAY` share common modules and extmodules.
- public modules `SHALL NOT` be instantiated by any other modules.

Reviewer 2: Oh my goodness, totally! Hahaha, thanks!

D'oh!

Thanks for the `MAY` / `SHALL NOT` language :+1: . Okay, great.  That makes a lot of sense.

Author: There's no requirement on disjointness. This statement was more to make it clear that the public modules _may_ have entirely disjoint instance graphs. However, they may also share anything. Put differently, circuits are collections of `public` modules. What those `public` modules instantiate or reference, has no real bearing on anybody using them as the rest of this isn't visible via the ABI.

In the circuit `Qux` below, there are two public modules, `Corge` and `Grault`.

> Neither instantiates each other:

Reviewer 2: ```suggestion
Neither instantiates the other:
```
? I don't know but "neither.. each other" seems strange.   It's likely I'm wrong, whatever you think is best :+1:.

Author: Made this example longer and this language is no longer present. Good catch.

## Circuits

A FIRRTL circuit is a named collection of FIRRTL modules.  Each module is a
hardware "unit" that has ports, registers, wires, and may instantiate other
modules (see: [@sec:modules]).  (This is the same concept as a Verilog

> `module`{.verilog}.)  A public module is usable outside of the current circuit.

Reviewer: ```suggestion
`module`{.verilog}.)  A public module, when lowered, will be usable outside of the context of the current circuit.
```

Reviewer: i realized upon re-reading this that we are talking about usability at the verilog level, unless we also mean usability at a hypothetical firrtl-linking level, which i think is not what we are trying to say.

Author: I think we're punting on FIRRTL linking and saying that this always happens at the Verilog level.  This would be analogous to a C++ header file being enough to determine the name of a function that will be compiled by another process.  There's no C++ "linking" per-se.  However, for this to work there needs to be a paired FIRRTL `extmodule` that has the same name and ports.  (Which is the same as the C++ case.)

> `module`{.verilog}.)  A public module is usable outside of the current circuit.

Reviewer 3: Perhaps "Public modules may be instantiated outside the circuit".  This defined "usable" more narrowly.

Further, something about public modules being the exported identifiers for a circuit and all other module kinds not being namable outside the circuit might help clarify.

Consider the following circuit named `Foo`.  This contains two modules, `Bar`
and `Baz`.  Module `Baz` is marked public.

``` firrtl
circuit Foo :
  module Bar :
    input a: UInt<1>
    output b: UInt<1>

    connect b, a

  public module Baz :
    input a: UInt<1>
    output b: UInt<1>

    inst bar of Bar
    connect bar.a, a
    connect b, bar.b
```

A circuit may or may not contain a module with the same name as the circuit.  A
circuit may contain public modules that instantiate no common instances.  Two

> public modules may share common instances.

Author: I'll just drop this. The second example should be explanation enough.

> public modules may share common instances.

Author: Fully removed in 5676d11.

Circuit `Foo`, below, shows an example of a more complex circuit.  This circuit
contains three public modules, `Foo`, `Bar`, and `Baz`, and one non-public
module, `Qux`.  Module `Foo` has no common instances with `Bar` or `Baz`.
`Bar` and `Baz` both instantiate `Qux`:

``` firrtl
circuit Foo :
  public module Foo :

  public module Bar :
    inst qux of Qux

  public module Baz :
    inst qux of Qux

  module Qux :
```

A circuit that contains no public modules is trivially equivalent to a circuit
that contains no modules.  It is not enforced that a circuit has at least one

> public module, but it is expected that it does.

Reviewer 3: Is the empty circuit legal in firrtl currently?

I think I'm find with the empty circuit being legal.

instance statement for details on how to instantiate a module
([@sec:instances]).

### Public Modules

A public module is a module that is marked with the `public`{.firrtl} keyword.
This means that this is a module that is usable external to this circuit.  A
public module thereby has a defined ABI---knowing the FIRRTL representation is
sufficient to know its compiled representation.

A public module has a number of restrictions:

> 1. A public module may not be instantiated in the current circuit.

Reviewer: because of this point 1, i feel like instead of calling this "public", 'root' would be a better word. That captures more the meaning. We may decide that public is applicable to things that do not meet this requirement 1, later.

Author: We can just remove this requirement. I think Will, you, and others have all convinced me that this isn't useful to enforce. I'm planning to back this out and then add language around the types of optimizations that are allowed here. Specifically, a public module that is instantiated must be compiled as if it were not instantiated. However, a module that instantiates a public module is allowed to optimize through a public module.

E.g., it is legal for a parent to inline a public module that it instantiates. However, that public module must still be produced following the Verilog ABI.

Initial circuit:
```
circuit Foo:
  public module Foo:
    inst bar of Bar

  public module Bar:
    wire w: UInt<1>
```

Final legal circuit:

```
circuit Foo:
  public module Foo:
    wire bar_w: UInt<1>
  public module Bar:
    wire w: UInt<1>
```

E.g., if a parent instantiates a public module and drives a port to a constant that is fed out through an output port, then the parent module can constant prop through the instance. However, the constant cannot be moved into the instance.

Initial circuit:

```
circuit Foo:
  public module Foo:
    output b: UInt<1>
    inst bar of bar
    connect bar.a, UInt<1>(1)
    connect b, bar.b
  public module Bar:
    input a: UInt<1>
    output b: UInt<1>
    connect b, a
```

Optimized circuit:
```
  public module Foo:
    output b: UInt<1>
    connect b, UInt<1>(1)
  public module Bar:
    input a: UInt<1>
    output b: UInt<1>
    connect b, a
```

Illegally optimized circuit:
```
  public module Foo:
    output b: UInt<1>
    connect b, UInt<1>(1)
  public module Bar:
```

Reviewer: Ah, if you don't enforce it then all @dtzSiFive questions I said didn't matter start mattering again ;) But i think that the rule that a public module has to exist in a verilog file of that name squashes many of the questions.

Reviewer: > Specifically, a public module that is instantiated must be compiled as if it were not instantiated. However, a module that instantiates a public module is allowed to optimize through a public module.

Putting this in my own words:
- Optimizations on a module A that instantiates a public module B MAY treat B as if it were equivalent to an extmodule, but optimizations to A MAY also fully consider the contents of B and perform any modifications to A desired (including not instantiating B at all).

- When a module A instantiates a public module B, optimizations on B must ignore any information known about its context in A

Reviewer: i am a little bit worried about the first bullet, sicne it might imply that you are actually essentially duplicating the public module into a "not public" copy and using that, which might not get the desired effect.

Reviewer: in the future we could add a thing at the instantiation site that is `inst foo of Foo as public` which means "use the public version of this and I mean it" to block the optimizations the other way.

Reviewer 2: Spec currently contains this language:

> Modules have the property that instances can always be *inlined* into the parent
module without affecting the semantics of the circuit.

Inlining I think usually means replacing a thing with its definition, not necessarily deleting the original.  Often the latter is kinda assumed since often after inlining it everywhere the definition is now dead.  But as-is I'd read this to mean it's fine to indeed copy any module's definition out (?) to replace an instance.  With or without public -- but with public it'll never be dead so just want to call this out a bit.

You're absolutely right that the first bullet, by allowing optimizing A using B's definition is essentially equivalent to allowing (at least in some senses) inlining B's definition entirely. :+1:

Maybe we need some way to express/manage that the "identity" (and count) of certain things matter, as well as perhaps their position in some way.  For example, probably don't want to be cloning registers around? But dunno.

Pinning this down would help a lot I think, and has been run into on other related fronts.

Anyway good points, personally I think this should be allowed (copying B into A) and if it's not we should figure out why precisely and pin that down.

> 1. A public module may not be instantiated in the current circuit.

Reviewer: would it be acceptable to instantiate a public aka root module *as an ext module (would need a new API)* in the current circuit? If not, why not?

Reviewer 2: :exploding_head: good question / fun idea!

1. A public module may have no ports of uninferred width.

> 1. A public module may have no ports of abstract reset type.

Reviewer 2: FWIW also no input probes, can cannot RWProbe a public module's ports.  This is in the text elsewhere already, and may not need to be repeated here.

instance statement for details on how to instantiate a module
([@sec:instances]).

### Public Modules

> A public module is a module that is marked with the `public`{.firrtl} keyword.
> This means that this is a module that is usable external to this circuit.  A
> public module thereby has a defined ABI---knowing the FIRRTL representation is
> sufficient to know its compiled representation.

Reviewer: ```suggestion
A public module is a module that is marked with the `public`{.firrtl} keyword.
A public module has a defined ABI---knowing the FIRRTL representation is
sufficient to know its compiled representation.
This means that this is a module that is usable external to this circuit.  
```

Suggest switching the order of these things -- it's usable *because* it has a defined ABI.

Author: I'm also going to switch to saying "available" and not "usable" as the former language shows up in the LLVM linkage type that Will referenced.

instance statement for details on how to instantiate a module
([@sec:instances]).

### Public Modules

A public module is a module that is marked with the `public`{.firrtl} keyword.
This means that this is a module that is usable external to this circuit.  A
public module thereby has a defined ABI---knowing the FIRRTL representation is
sufficient to know its compiled representation.

The example below shows a public module `Foo`:

``` firrtl
public module Foo:
  input a: UInt<1>
  output b: UInt<1>

  connect b, a
```

A public module has a number of restrictions:

1. A public module may have no ports of uninferred width.
1. A public module may have no ports of abstract reset type.
1. A public module may have no ports of input probe type.

> 1. A `RWProbe`{.firrtl} may not be used to access a public module's ports.

Reviewer: at some point we're going to need to relax this RWProbe thing...

Reviewer 2: In short, this is a rule because of the net propagation problems that violate basic FIRRTL semantics and cause problems in practice that can be a pain to track down (and are not consistently simulated).

So IMO it's not about relaxing it's about not breaking things that need to work.  What motivates this comment / what's the ask?

Since you brought it up :wink:.... none of this is new but perhaps needs reviving:

As long as we don't emit ports as variables (logic), and as nets (as we do now), I think we want that restriction.  I'm not sure how reasonable or not it'd be to move forward with such a change.  Do you have a sense of that?
I think this was discussed long ago and maybe even attempted but I don't know the details.

One idea: require as part of the extmodule/module signature, and ABI, that it may be forced and undefined behavior to have it wrong re:forceability on either side of the boundary.  This could in practice be "emit as variable", but selectively not everywhere (in trade for requiring correct declarations or bugs), basically encode the forceability on the type itself (not the rwprobe, the target port).  This doesn't really feel like a win, but maybe others have other opinions.

Or: we let the net propagation problems be users problem/implementation (/simulator-)defined behavior, hopefully giving them a way to mitigate that isn't manually inserting buffer wires and marking don't touch in order to use this feature or in defense of possibility it's used in a module they don't own (this happens in practice in designs "in the wild" due to adhoc use of force XMR's).  This has benefit of not getting in the way of users who don't want this restriction, but is my least favorite as I'd rather make RWProbe a useful robust concept that acts as folks might expect it to.

WDYT? Is there a way to get what drives this ask a different way?

Basically it's for those reasons it's a requirement, no disagreement that it'd be better to not have such a rule if there's a better place we can take things :+1: .

Reviewer 2: Anyway this probably should be discussed elsewhere, but don't mean to close this down -- please LMK here or elsewhere if have any further thoughts/ideas/suggestions.

Reviewer: Ah, I misread this as "there may not be any RWProbe type ports". Which is not in fact what this says, and now I understand your objection.

instance statement for details on how to instantiate a module
([@sec:instances]).

### Public Modules

A public module is a module that is marked with the `public`{.firrtl} keyword.
A public module has a defined ABI---knowing the FIRRTL representation is
sufficient to know its compiled representation.  This means that a public module

> is available external to this circuit.

Reviewer 3: I don't think this strictly follows.  I could know the representation, but migrate all the contents to the in-circuit instantiators.  Public requires stronger semantics.

Reviewer 3: This is in fact stated later, which makes this partially redundant, but in a way which emphasizes just a part as though that was the most important thing.

Author: I'll drop this and switch to simpler language about public being capable of being instantiated externally.

The example below shows a public module `Foo`:

``` firrtl
public module Foo:
  input a: UInt<1>
  output b: UInt<1>

  connect b, a
```

A public module has a number of restrictions:

1. A public module may have no ports of uninferred width.
1. A public module may have no ports of abstract reset type.

> 1. A public module may have no ports of input probe type.

Reviewer 3: This pains me, but I don't see a way around it.

> 1. A `RWProbe`{.firrtl} may not be used to access a public module's ports.

Reviewer 2: If we clarify the difference, that would help yes, and related issue of whether either is expected to actually force the ports themselves necessarily or only have in some sense equivalent behavior.

This restriction (currently in the spec, just described here now) is to ensure a FIRRTL compiler can emit what's needed on /both/ sides of a module with forced ports, to give room for bounce wires or whatever, since "may be forced" is not part of the signature/ABI.

Emitting ports (and other signals) as variables would mostly eliminate the need for this restriction, as you mention, and we should clarify the difference between rwprobe'ing instance results vs the ports from inside the module so that on, e.g., extmodule's that may not have ports as variables, there's no surprises.

Sound good?

> 1. A `RWProbe`{.firrtl} may not be used to access a public module's ports.

Reviewer 4: What constitutes a "use"?

Is there an intuition for why this rule is in place?

1. A public module's ports are lowered using one of the port lowering ABIs
   defined in the FIRRTL ABI specification.

> 1. A public module may be instantiated by other modules, but must be compiled

Reviewer 3: Alternately: "A public module may be instantiated by other modules within a circuit, but the behavior of the module must not be affected by these instantiations."

instance statement for details on how to instantiate a module
([@sec:instances]).

### Public Modules

A public module is a module that is marked with the `public`{.firrtl} keyword.
A public module may be instantiated outside by other circuits.

The example below shows a public module `Foo`:

``` firrtl
public module Foo:
  input a: UInt<1>
  output b: UInt<1>

  connect b, a
```

A public module has a number of restrictions:

1. A public module may have no ports of uninferred width.
1. A public module may have no ports of abstract reset type.
1. A public module may have no ports of input probe type.
1. A `RWProbe`{.firrtl} may not be used to access a public module's ports.
1. A public module's ports are lowered using one of the port lowering ABIs

>    defined in the FIRRTL ABI specification.

Reviewer 4: This doesn't seem like it should be a "restriction". This is a commentary on this proposal's relation to the ABI. 

We should scootch it down a few lines, instead of having it appear in this list. And I haven't read the latest version of the ABI spec, but when this says "one of the port lowering ABIs", the question that comes to my mind is, "which one? And who decides?"

Reviewer 4: Also, remove "The last point requires clarification" from the next sentence. This is just evidence that commentary on ABI is a piece of commentary and not a restriction :)

Author: Agree. This isn't a restriction. I'll drop it. You're question about which one/who decides is important and the current spec is silent on this. The only way I know of to properly fix this is to add the port lowering convention to each external module and any public module. For now, the spec can choose to be silent about this. However, we should add this.

Will drop the "requires clarification" sentence.

Author: Fixed in 90f91c0.

1. A public module may be instantiated by other modules within a circuit, but
   the behavior of the module must not be affected by these instantiations.

> The last point requires clarification.  A public module may or may not be

Reviewer 4: Let's not use the phrase "may or may not" here, since either one alone sounds like a prescription.

Reading this closer, this whole paragraph is confusing to me. What does it mean to be "compiled" with respect to the rest of the spec? It comes off as if it can be instantiated without being instantiated :)

Author: The problem here is that there is no good way of indicating commentary in the spec. The RISC-V specifications have shaded blocks where they do this and this may be useful to add to the FIRRTL spec. 

None of the body of this paragraph has any bearing on the spec and is only providing commentary on how to interpret the last item (line 246) if you are building a FIRRTL compiler. If you need to produce Verilog from a public module, this will make sense to you. If you are trying to understand FIRRTL, then it has no meaning.

I'll move some of this language into the ABI doc as that is more appropriate.

``` firrtl
circuit Top : @[myfile.txt 14:8]
  public module Top : @[myfile.txt 15:2]
    output out: UInt @[myfile.txt 16:3]
    input b: UInt<32> @[myfile.txt 17:3]

>     input c: UInt<1> @[myfile.txt 18:3]

Reviewer: i think we need to add the `public` keyword to the grammar?

And release notes
